yaks/lib/yaks/default_policy.rb

```diff
@@ -17,13 +19,59 @@ def initialize(options = {})
       @options = DEFAULTS.merge(options)
     end
 
+    # Main point of entry for mapper derivation. Calls
+    # derive_mapper_from_collection or derive_mapper_from_item
+    # depending on the model.
+    #
     # @param model [Object]
     # @return [Class] A mapper, typically a subclass of Yaks::Mapper
     #
-    # @raise [NameError] only occurs when the model is anything but a collection.
+    # @raise [RuntimeError] occurs when no mapper is found or a collection
+    #                       is given but does not respond to :first
     def derive_mapper_from_object(model)
+      mapper = detect_configured_mapper_for(model)
+      return mapper if mapper
       return derive_mapper_from_collection(model) if model.respond_to? :to_ary
-      derive_mapper_from_single_object(model)
+      derive_mapper_from_item(model)
+    end
+
+    # Derives a mapper from the given collection.
+    #
+    # @param collection [Object]
+    # @return [Class] A mapper, typically a subclass of Yaks::Mapper
+    #
+    # @raise [RuntimeError] when collection does not respond to :first
+    def derive_mapper_from_collection(collection)
+      raise_not_a_collection(collection) unless collection.respond_to?(:first)
```

Comment:
> What if the user calls this method directly and gives it an object that is not a collection?

That's not a use case we have to worry about. Overriding it is part of the public interface, using it for some other use is not.

About `#to_ary`, if an object implements this it implies that it's indistinguishable from an array, as opposed to `#to_a` which just means there is some conceivable way to turn it into an array. This is called implicit vs explicit conversion protocols. So an object that implements `to_ary` but not `first` would IMO be a violation of this protocol.

On the other hand you never know what people come up with. What Ruby itself usually does in  these cases is check if `to_ary` (or `to_str`, etc) is implemented, then call `to_ary` and continue with the result assuming it will behave like an array. so we probably should do the same.